app/src/behaviors/behavior_hold_tap.c

```diff
@@ -0,0 +1,491 @@
+		if (captured_events[i] == NULL) {
+			captured_events[i] = event;
+			return 0;
+		}
+	}
+	return -ENOMEM;
+}
+
+static struct position_state_changed *find_captured_keydown_event(u32_t position)
+{
+	struct position_state_changed *last_match = NULL;
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS; i++) {
+		const struct zmk_event_header *eh = captured_events[i];
+		if (eh == NULL) {
+			return last_match;
+		}
+		if (!is_position_state_changed(eh)) {
+			continue;
+		}
+		struct position_state_changed *position_event = cast_position_state_changed(eh);
+		if (position_event->position == position && position_event->state) {
+			last_match = position_event;
+		}
+	}
+	return last_match;
+}
+
+static void release_captured_events()
+{
+	if (undecided_hold_tap != NULL) {
+		return;
+	}
+
+	// We use a trick to prevent copying the captured_events array.
+	//
+	// Events for different mod-tap instances are separated by a NULL pointer.
+	//
+	// The first event popped will never be catched by the next active hold-tap
+	// because to start capturing a mod-tap-key-down event must first completely
+	// go through the events queue.
+	//
+	// Example of this release process;
+	// [mt2_down, k1_down, k1_up, mt2_up, null, ...]
+	//  ^
+	// mt2_down position event isn't captured because no hold-tap is active.
+	// mt2_down behavior event is handled, now we have an undecided hold-tap
+	// [null, k1_down, k1_up, mt2_up, null, ...]
+	//        ^
+	// k1_down  is captured by the mt2 mod-tap
+	// !note that searches for find_captured_keydown_event by the mt2 behavior will stop at the first null encountered
+	// [mt1_down, null, k1_up, mt2_up, null, ...]
+	//                  ^
+	// k1_up event is captured by the new hold-tap:
+	// [k1_down, k1_up, null, mt2_up, null, ...]
+	//                        ^
+	// mt2_up event is not captured but causes release of mt2 behavior
+	// [k1_down, k1_up, null, null, null, ...]
+	// now mt2 will start releasing it's own captured positions.
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_CAPTURED_EVENTS; i++) {
+		const struct zmk_event_header *captured_event = captured_events[i];
+		if (captured_event == NULL) {
+			return;
+		}
+		captured_events[i] = NULL;
+		if (undecided_hold_tap != NULL) {
+			k_msleep(10);
+		}
+		if (is_position_state_changed(captured_event)) {
+			struct position_state_changed *position_event = cast_position_state_changed(captured_event);
+			LOG_DBG("Releasing key position event for position %d %s", position_event->position, (position_event->state ? "pressed" : "released"));
+		} else {
+			struct keycode_state_changed *modifier_event = cast_keycode_state_changed(captured_event);
+			LOG_DBG("Releasing mods changed event 0x%02X %s", modifier_event->keycode, (modifier_event->state ? "pressed" : "released"));
+		}
+		ZMK_EVENT_RELEASE_AGAIN(captured_event);
+	}
+}
+
+
+/************************************************************    ACTIVE TAP HOLD HELPER FUNCTIONS */
```

Comment:
I don't love this convention for comments for sections.

Convince me?

Reply:
Not a big fan either, but I noticed I was getting lost in the file. The file is pushing 500 lines which is about the max I'm comfortable with. I'm not sure if/how to cut it up. 

For now I've removed the sections, I'm open to suggestions.

Reply:
Let's ponder. Also, lets not hold this PR up on that. Thanks.